File: src/day06.js
fs = require("fs");


function updateDepths(node) {
    node.children.forEach(child => {
        child.depth = node.depth + 1;
        updateDepths(child);
    });
}

function buildOrbits() {
    const file = fs.readFileSync("src/day06.input", "utf8").split("\r\n");
    let orbits = {};
    let center = null;
    let you = null;
    let san = null;
    file.forEach(element => {
        const [lhs, rhs] = element.split(")");
        // console.log("mapping", lhs, rhs);
        let parent = orbits[lhs];
        if (!parent) {
            orbits[lhs] = {
                name: lhs,
                depth: 0,
                children: []
            }

            parent = orbits[lhs];
        }

        let child = orbits[rhs];
        if (!child) {
            orbits[rhs] = {
                name: rhs,
                children: []
            }

            child = orbits[rhs];
        }

        if (parent.name === "COM") {
            center = parent;
        }

        if (child.name === "YOU") {
            you = child;
        } else if (child.name === "SAN") {
            san = child;
        }

        parent.children.push(child);
        child.parent = parent;
        child.depth = parent.depth + 1;
        updateDepths(child);
    });

    return {
        center,
        you,
        san
    }
}


function traverse(node, counter) {
    node.children.forEach(child => {
        counter += traverse(child, 0);
        counter += child.depth;

        // console.log("visiting", child.name, counter);
    });

    return counter;
}

function findCommonAncestor(you, san) {
    console.log("Finding ancestors for", you, san);

    const yourAncestors = [you.parent];
    let node = you.parent;
    while (node.parent) {
        yourAncestors.unshift(node.parent.name);
        node = node.parent;
    }

    const santaAncestors = [san.parent];
    node = san.parent;
    while (node.parent) {
        santaAncestors.unshift(node.parent.name);
        node = node.parent;
    }

    console.log(yourAncestors, santaAncestors);

    while (yourAncestors[0] === santaAncestors[0]) {
        yourAncestors.shift();
        santaAncestors.shift();
    }

    console.log("distance", yourAncestors.length + santaAncestors.length);
}


const data = buildOrbits();
total = traverse(data.center, 0);
console.log("total", total);

findCommonAncestor(data.you, data.san);

File: src/day06.rs
use std::collections::HashMap;
use std::fs;

struct Node {
    name: String,
    parent: Option<String>,
    children: Vec<Node>,
    depth: i32,
}

impl PartialEq for Node {
    fn eq(&self, other: &Self) -> bool {
        self.name == other.name
    }
}

pub fn first() {
    let content = fs::read_to_string("./src/day06.input").expect("error reading file");
    let split = content.split("\n");
    let mut orbits = HashMap::new();
    for line in split {
        let parts: Vec<&str> = line.split(")").collect();
        let parent_name = parts[0].to_string();
        let child_name = parts[1].to_string();
        let parent = orbits.get(&parent_name);
        if parent == None {
            let parent_node = Node {
                name: parent_name.clone(),
                parent: None,
                children: Vec::new(),
                depth: 0,
            };
            orbits.insert(parent_name.clone(), &parent_node);
        }

        let child = orbits.get(&child_name);
        if child == None {
            let child_node = Node {
                name: child_name.clone(),
                parent: Some(parent_name.clone()),
                children: Vec::new(),
                depth: 0,
            };
            orbits.insert(child_name.clone(), &child_node);
        }

        // Now both nodes exist -> update depth
        let parent_node = orbits.get(&parent_name).unwrap();
        let child_node = orbits.get(&child_name).unwrap();
        parent_node.children.push(*child_node);

        // ... I give up!
    }
}
